File: 2019/C++/src/day1.cpp
/**
 * \file day1.cpp
 * \brief AoC 2019 - Day 1 (C++ version)
 * \author Mina Pêcheux
 * \date 2019
 *
 * [ ADVENT OF CODE ] (https://adventofcode.com)
 * ---------------------------------------------
 * Day 1: The Tyranny of the Rocket Equation
 * =============================================
 */
#include "utils.hpp"
#include "parser.hpp"

// [ Computation functions ]
// -------------------------

/**
 * \fn int computeFuel(int mass)
 * \brief Computes the required fuel for a module of given mass.
 *
 * \param mass The mass of the module to compute the fuel consumption for.
 * \return Required amount of fuel.
 */
int computeFuel(int mass) {
  return mass / 3 - 2;
}

/**
 * \fn int computeTotalFuel(int mass)
 * \brief Computes the total required fuel for a module of given mass and the
 * added fuel, and so on. It works recursively until the computed amount of
 * fuel is zero or negative.
 *
 * \param mass The mass of the module to compute the fuel consumption for.
 * \return Required amount of fuel.
 */
int computeTotalFuel(int mass) {
  int f = computeFuel(mass);
  return (f <= 0) ? 0 : f + computeTotalFuel(f);
}

// [ Base tests ]
// --------------

/**
 * \fn void makeTests()
 * \brief Performs tests on the provided examples to check the result of the
 * computation functions is ok.
 */
void makeTests() {
  // Part I
  assert(computeFuel(12) == 2);
  assert(computeFuel(14) == 2);
  assert(computeFuel(1969) == 654);
  assert(computeFuel(100756) == 33583);
  // Part II
  assert(computeTotalFuel(14) == 2);
  assert(computeTotalFuel(1969) == 966);
  assert(computeTotalFuel(100756) == 50346);
}

int main(int argc, char const *argv[]) {
  // check function results on example cases
  makeTests();
  
  // get input data
  std::string dataPath = "../data/day1.txt";
  std::string data = readFile(dataPath);
  std::vector<int> inputs = parseToIntsWithDelimiter(data, "\n");
  
  // Part I
  int solution1 = 0;
  for (auto i : inputs) {
    solution1 += computeFuel(i);
  }
  std::cout << "PART I: solution = " << solution1 << '\n';
  
  // Part II
  int solution2 = 0;
  for (auto i : inputs) {
    solution2 += computeTotalFuel(i);
  }
  std::cout << "PART II: solution = " << solution2 << '\n';
  
  return 0;
}

File: 2019/CSharp/day1.cs
/*
 * =============================================
 * [ ADVENT OF CODE ] (https://adventofcode.com)
 * 2019 - Mina Pêcheux: C# version
 * ---------------------------------------------
 * Day 1: The Tyranny of the Rocket Equation
 * ============================================= */

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;

namespace CSharp2019
{
    class Day1
    {
        // [ Input parsing functions ]
        // ---------------------------

        /// <summary>Parses the incoming data into processable inputs.</summary>
        /// <param name="lines">Provided problem data.</param>
        /// <returns>Parsed data as a list of ints.</returns>
        static int[] ParseInput(IEnumerable<string> lines)
        {
            return lines
                .Where(l => l.Length > 0)
                .Select(l => int.Parse(l))
                .ToArray();
        }

        // [ Computation functions ]
        // -------------------------
        // == PART I

        /// <summary>Computes the required fuel for a module of given mass.</summary>
        /// <param name="moduleMass">The mass of the module to compute the fuel consumption for.</param>
        /// <returns>Required amount of fuel.</returns>
        static int ComputeFuel(int moduleMass)
        {
            return (int) Math.Floor(moduleMass / 3.0f) - 2;
        }

        // == PART II

        /// <summary>Computes the total required fuel for a module of given mass and the
        /// added fuel, and so on. It works recursively until the computed amount of
        /// fuel is zero or negative.</summary>
        /// <param name="moduleMass">The mass of the module to compute the fuel consumption for.</param>
        /// <returns>Required amount of fuel.</returns>
        static int ComputeTotalFuel(int moduleMass)
        {
            int fuel = ComputeFuel(moduleMass);
            return fuel <= 0 ? 0 : fuel + ComputeTotalFuel(fuel);
        }

        // [ Base tests ]
        // --------------
        static void MakeTests()
        {
            // == PART I
            Debug.Assert(ComputeFuel(12) == 2);
            Debug.Assert(ComputeFuel(14) == 2);
            Debug.Assert(ComputeFuel(1969) == 654);
            Debug.Assert(ComputeFuel(100756) == 33583);

            // == PART II
            Debug.Assert(ComputeTotalFuel(14) == 2);
            Debug.Assert(ComputeTotalFuel(1969) == 966);
            Debug.Assert(ComputeTotalFuel(100756) == 50346);
        }

        static void Main(string[] args)
        {
            // check function results on example cases
            MakeTests();

            // get input data
            string dataPath = "../data/day1.txt";
            int[] inputs = ParseInput(File.ReadLines(dataPath));

            // == PART I
            int solution = inputs.Select(i => ComputeFuel(i)).Sum();
            Console.WriteLine($"PART I: solution = {solution}");

            // == PART II
            solution = inputs.Select(i => ComputeTotalFuel(i)).Sum();
            Console.WriteLine($"PART II: solution = {solution}");
        }
    }
}
